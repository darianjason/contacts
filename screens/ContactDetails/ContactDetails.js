import React, {useLayoutEffect} from 'react';
import {ScrollView, View, Dimensions} from 'react-native';

import {DefaultText, Button, ProfilePicture} from '../../components/ui';
import styles from './ContactDetails.styles';

const setNavOptions = (navigation, id, fullName) => {
  navigation.setOptions({
    headerTitle: fullName,
    headerRight: () => (
      <Button
        name="edit"
        onPress={() => navigation.navigate('ManageContact', {id})}
      />
    ),
  });
};

const ContactDetails = ({navigation, route}) => {
  const {contact} = route.params;

  const {id, photo, firstName, lastName, age} = contact;
  const fullName = `${firstName} ${lastName}`;

  useLayoutEffect(() => {
    setNavOptions(navigation, id, fullName);
  }, [navigation, id, fullName]);

  return (
    <ScrollView contentContainerStyle={styles.screen}>
      <View>
        <ProfilePicture
          photo={photo}
          firstName={firstName}
          lastName={lastName}
          pictureSize={Dimensions.get('window').height / 4}
          initialsSize={36}
        />
        <DefaultText style={styles.name}>{fullName}</DefaultText>
      </View>

      <View style={styles.infoContainer}>
        <DefaultText>{age} years old</DefaultText>
      </View>
    </ScrollView>
  );
};

export default ContactDetails;
