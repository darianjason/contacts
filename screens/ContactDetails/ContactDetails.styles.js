import {StyleSheet} from 'react-native';

import {Colors} from '../../constants';

const styles = StyleSheet.create({
  screen: {
    flex: 1,
    alignItems: 'center',
    padding: 36,
  },
  name: {
    fontFamily: 'Manrope-Bold',
    fontSize: 24,
    marginTop: 16,
    alignSelf: 'center',
    color: Colors.primary,
  },
  infoContainer: {
    marginTop: 16,
  },
});

export default styles;
