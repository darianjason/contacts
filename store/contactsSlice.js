import {createSlice} from '@reduxjs/toolkit';

const contactsSlice = createSlice({
  name: 'contacts',
  initialState: {
    contacts: [],
  },
  reducers: {
    setContacts: (state, action) => {
      state.contacts = action.payload;
    },
    insertContact: (state, action) => {
      state.contacts.push(action.payload);
    },
    updateContact: (state, action) => {
      const {id, data} = action.payload;
      const updatedContactIndex = state.contacts.findIndex(
        contact => contact.id === id,
      );
      state.contacts[updatedContactIndex] = data;
    },
    deleteContact: (state, action) => {
      const {id} = action.payload;
      const index = state.contacts.findIndex(contact => contact.id === id);
      state.contacts.splice(index, 1);
    },
  },
});

const {actions, reducer} = contactsSlice;
export const {setContacts, insertContact, updateContact, deleteContact} =
  actions;
export default reducer;

const BASE_URL = 'https://simple-contact-crud.herokuapp.com';

export const fetchContacts = () => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact`);
    const contacts = await response.json();
    dispatch(setContacts(contacts.data));
  } catch (error) {
    throw error;
  }
};

export const addContact = contact => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(contact),
    });

    const responseJson = await response.json();
    console.log(responseJson);

    dispatch(insertContact(contact));
  } catch (error) {
    throw error;
  }
};

export const editContact = (id, data) => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    });

    const responseJson = await response.json();
    console.log(responseJson);

    dispatch(updateContact(id, data));
  } catch (error) {
    throw error;
  }
};

export const removeContact = id => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact/${id}`, {
      method: 'DELETE',
    });

    const responseJson = await response.json();
    console.log(responseJson);

    dispatch(deleteContact(id));
  } catch (error) {
    throw error;
  }
};
