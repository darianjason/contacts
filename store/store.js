// @flow

import {configureStore} from '@reduxjs/toolkit';

import contactsReducer from './contactsSlice';

type Contact = {
  id?: string,
  firstName: string,
  lastName: string,
  age: number | string,
  photo: string,
};

type Store = {
  contacts: Array<Contact>,
};

export const store: Store = configureStore({
  reducer: {
    contacts: contactsReducer,
  },
});
