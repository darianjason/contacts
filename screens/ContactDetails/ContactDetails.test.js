import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';

import {store} from '../../store/store';
import ContactDetails from './ContactDetails';

describe('Contact Details screen', () => {
  test('renders correctly', () => {
    const navigation = {setOptions: jest.fn()};
    const route = {
      params: {
        contact: {
          id: 'test',
          firstName: 'Hello',
          lastName: 'World',
          age: '99',
          photo: 'N/A',
        },
      },
    };

    const tree = renderer.create(
      <Provider store={store}>
        <ContactDetails navigation={navigation} route={route} />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });
});
