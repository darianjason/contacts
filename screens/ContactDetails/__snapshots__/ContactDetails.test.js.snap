// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact Details screen renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
      "flex": 1,
      "padding": 36,
    }
  }
>
  <View>
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#E2EEF3",
            "borderRadius": 8,
            "height": 333.5,
            "justifyContent": "center",
            "overflow": "hidden",
            "width": 333.5,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#142B3E",
              "fontFamily": "Manrope-Bold",
              "fontSize": 36,
              "letterSpacing": 2,
            }
          }
        >
          H
          W
        </Text>
      </View>
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#142B3E",
            "fontFamily": "Manrope-Bold",
            "fontSize": 24,
            "marginTop": 16,
          }
        }
      >
        Hello World
      </Text>
    </View>
    <View
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#060A0E",
            "fontFamily": "Manrope-Regular",
          }
        }
      >
        99
         years old
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
